_posts/2019-01-02-Deviceid-Reliability.md

```diff
@@ -0,0 +1,95 @@
+The peaks in at one are those which are visited by always distinct deviceid while the other agglomerate close to zero.
+
+We suspect that something is broken in the way the apps provide/access the deviceID information for those apps.
+All deviceIDs which appear in a non reliable flagged app are considered non reliable, a simple inner join over the selected apps is enought to create the blacklist.
+
+
+* **Non realistic number of apps**<br/>
+For a fixed amount of time (e.g. 1 day) it is possible to observe the number of apps visited by each deviceID.
+From this plot we found out some clear outliers which cannot correspond to real human users.
+![num_uniqu_apps](../images/Deviceid-Reliability/num_unique_apps.png)
+In the plot the number of the vistited apps is standardized (the absolute value minus the overall mean divided by the standard deviation), so that the presence of outliers is evident.
+
+
+* **Short period of time**:<br/>
+When we observe a deviceID on our data for a very short period of time (say 1 minute) but we do not register any other occurrence of the deviceID over several days both in the past and in the future, we do not consider these deviceIDs to be worth of further analysis.
+In fact, it is not sure that we have a real person behind it, for example a machine which automatically reset the value would produce the same data.
+![one_over_more_days](../images/Deviceid-Reliability/only_1_observations_over_days.png)
+The plot above represents the observation of a deviceID in a 5 days period, both for users which shown a regular activity (d6, d7, d8) and for those which appear only once (d1, d2, d3).
+The y axis is used to facilitate the graphical visualization of the activity patterns.
+
+
+* **Randomly generated deviceID**:<br/>
+Some app could provide a randomly generated deviceID.
+If this is the case, it is expected that the same deviceID is not observed in any other apps.<br/>
+We can think about two ways to generate the values:
+1) a new random deviceID is created for every acces;
+2) a particular transformation of the deviceID is used, thus the provided deviceID for the same real device is always the same but it will not be allined with the one provided by other apps.<br/>
+The query below allows us to catch these apps:
+```sql
+SELECT app, COUNT(deviceid) as number_of_distinct_deviceid
+FROM
+    (
+    SELECT deviceid, COUNT(DISTINCT apps) as num_apps_visited
+    FROM table_apps_deviceid_visits
+    GROUP BY deviceid
+    ) t1
+JOIN
+    (
+    SELECT deviceid, app
+    FROM table_apps_deviceid_visits
+    ) t2
+ON t1.deviceid=t2.deviceid
+GROUP BY app
+HAVING MAX(num_apps_visited) == 1
+```
+Where the combination deviceID-app is unique in the table table_apps_deviceid_visits.
+Plotting the histogram of the number_of_distinct_deviceid we can recognize that certain apps have a huge numbers of unique devices, all these apps are considered to gives unreliable deviceID values.
```

Comment:
Hm, wouldn't a better plot be the fraction of shared device IDs. The ones we want to filter out should all be at zero.

Reply:
The plot was wrong. Now there is the one you wanted. In the end we did not use this cut in production since we did not found any devices such as illustrated in the histogram.
I rephrased the text to explain why we exclude. I think it is still interesting to show that we investigate this pattern and found nothing. However, I can understand that blinded reader could see this as a weak point instead of plus. I am open for suggestions if it is worth to keep this bullet point or not.